### PR TITLE
#167377493 Users should be able Signup/Register to 1kbIdeas

### PR DESCRIPTION
**Pull request by Jonathan4github**
### What does this PR do?
This PR implements a signup component
#### Description of Task to be completed
- Create a re-useable input textbox component
- Create signup actions and reducers
#### How should this be manually tested?
- Clone and cd into this repository
- run ```npm install```&```npm run start:dev```
- Navigate to http://localhost:{Port}/signup
- Then supply the user detail data and submit

#### Any Background Context You want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#167377493](https://pivotaltracker.com/story/show/167377493)

#### Screenshots (if appropriate)
Screenshot 2019-09-04 at 7 43 02 AM
Screenshot 2019-09-04 at 8 17 25 AM

### CODE REVIEW

**Comment by iverenshaguy** [comment]
I merged the conflicting PR since morning. I'm surprised it has not been rebased yet. Please what is going on?

**Comment by Jonathan4github** [comment]
> I merged the conflicting PR since morning. I'm surprised it has not been rebased yet. Please what is going on?

I have rebased the branch.

File: src/components/Signup/Signup.spec.jsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import jest from 'jest-mock';
```
Comment (Jonathan4github) [inline src/components/Signup/Signup.spec.jsx]: remove jest-mock

File: src/components/Signup/Signup.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import { SignupComponent } from './index';
+
+describe('Signup', () => {
+  it('should render with out error', () => {
+    const createUserRequest = jest.fn();
+    const message = 'message';
```
Comment (Jonathan4github) [inline src/components/Signup/Signup.spec.jsx]: auth props change to success message

**Comment by iverenshaguy** [comment]
@successgilli Please review this PR and ensure that it adheres to our convention, thanks

File: package.json [hunk 3]
```diff
@@ -41,14 +41,16 @@
     "debug": "^4.1.1",
     "dotenv": "^8.1.0",
     "express": "^4.17.1",
+    "jest-mock": "^24.9.0",
```
Comment (successgilli) [inline package.json]: Please, jest-mock is already integrated into jest.
You can uninstall this.

File: src/components/Signup/Signup.spec.jsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
```
Comment (successgilli) [inline src/components/Signup/Signup.spec.jsx:2]: Add the proper comments and import ordering. e.g // styles

File: src/components/Signup/index.jsx [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+    const userMessage = success ? 'check your email for the confirmation link!' : error;
+
+    return ((
+      <div>
+        <form className="form" onSubmit={this.handleSubmit}>
+          { showMessage && <div className={classList}>{userMessage}</div>}
+          <div className="section">
+            <label className="form__label" htmlFor="firstName">
+              <input
```
Comment (successgilli) [inline src/components/Signup/index.jsx]: Please, use the reusable input component in our project

**Review by successgilli — CHANGES_REQUESTED** [review]
Nice job so far @Jonathan4github.
Please, review my comments and reproduce them in all files affected.
Thanks

File: src/components/Signup/Signup.scss [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    background-color: #FFBABA;
+    color: #D8000C;
+  }
+
+  .form__label {
+    width: 80%;
+    margin: 0;
+    box-sizing: border-box;
+
+    .label__inputElem {
```
Comment (Jonathan4github) [inline src/components/Signup/Signup.scss]: remove camel case

File: src/components/Signup/Signup.scss [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      font-size: 1rem;
+      outline: 0;
+    }
+  }
+
+  .btn-section {
+    display: flex;
+    justify-content: flex-end;
+
+    .form__submitBtn {
```
Comment (Jonathan4github) [inline src/components/Signup/Signup.scss]: remove camel case

File: src/components/Signup/Signup.scss [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      border: 0;
+      border-radius: 0.4rem;
+      font-size: 1.25rem;
+      cursor: pointer;
+      color: #FFF;
+      outline: 0;
+    }
+  }
+
+  .loader-div {
```
Comment (Jonathan4github) [inline src/components/Signup/Signup.scss]: loader wrapper

File: src/components/Signup/Signup.scss [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      background-color: #FD8234;
+      width: 7.5rem;
+      height: 2.7rem;
+      padding: 0.6rem;
+      box-sizing: border-box;
+      border: 0;
+      border-radius: 0.4rem;
+      font-size: 1.25rem;
+      cursor: pointer;
+      color: #FFF;
```
Comment (Jonathan4github) [inline src/components/Signup/Signup.scss]: all color code should be  in variables

File: src/components/Signup/index.jsx [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    *
+    * @param {Object} props - React PropTypes
+    *
+    * @property {Function} signupUser - handles signup action by dispatching signup request action
+    * @property {Object}  signupUser - object containing signup user details
+    *
+    */
+  static propTypes = {
+    createUserRequest: PropTypes.func.isRequired,
+    signup: PropTypes.string.isRequired
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: use shape instead of string

Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: change signin  and login to auth

File: src/components/Signup/index.jsx [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+  /**
+    * Handles input into the form by updating state with values
+    * @method
+    *
+    * @param {Object} e - target object
+    *
+    * @return {void}
+    */
+  handleInput = (e) => {
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: change e to event

File: src/components/Signup/index.jsx [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    const { name, value } = e.target;
+    const { errors, ...fields } = this.state;
+    const newFields = { ...fields, [name]: value };
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: change fields to ...others

File: src/components/Signup/index.jsx [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+    const validatorErrors = new Validation(signupValidation.Signup).validate(newFields);
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: export the signup

File: src/components/Signup/index.jsx [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+    this.setState({ errors: { ...errors, [name]: validatorErrors[name] }, ...newFields });
+  }
+
+  formValid = (formErrors) => !Object.values(formErrors).filter((val) => !!val).length
+
+  handleSubmit = (e) => {
+    e.preventDefault();
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: change e to event

File: src/components/Signup/index.jsx [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  render() {
+    const {
+      firstName, lastName, userName, email, password, confirmPassword, errors
+    } = this.state;
+    const { showMessage } = this.state;
+    const { signup: { success, error, isRequesting } } = this.props;
+    const classList = classNames({
+      'info-msg': success,
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: remove 113

File: src/components/Signup/index.jsx [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+      'success-msg': success,
+      'error-msg': error
+    });
+
+    const userMessage = success ? 'check your email for the confirmation link!' : error;
+
+    return ((
+      <div>
+        <form className="form" onSubmit={this.handleSubmit}>
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: change form to signup class

File: src/components/Signup/index.jsx [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+
+          { showMessage && <div className={classList}>{userMessage}</div>}
+          <div className="section">
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: change the section classs to another name

File: src/components/Signup/index.jsx [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+          </div>
+          <div className="section">
+            <label className="form__label" htmlFor="userName">
+              <Input
+                type="text"
+                name="userName"
+                value={userName}
+                className="label__inputElem"
+                placeholder="Username"
+                onChange={this.handleInput}
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx:160]: Capitalize placeholder

File: src/components/Signup/index.jsx [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+                placeholder="Confirm Password"
+                onChange={this.handleInput}
+                required
+              />
+              {<span className="form__errors">{errors.confirmPassword}</span>}
+            </label>
+          </div>
+          <div className="btn-section">
+            {
+              (isRequesting === true) && <div className="loader-div"><div className="loader" /></div>
```
Comment (Jonathan4github) [inline src/components/Signup/index.jsx]: isRequesting
!isRequesting

fix to one line

File: src/store/modules/signUp/index.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+import { call, put, takeLatest } from 'redux-saga/effects';
+
+// types
+import { CREATE_USER_REQUEST, SIGNUP_SUCCESS, SIGNUP_FAILURE } from 'modules/signUp/types';
+
+// requests
+import API from 'modules/signUp/requests';
+
+// actions
+export const createUserRequest = (bool) => ({ type: CREATE_USER_REQUEST, payload: bool });
```
Comment (Jonathan4github) [inline src/store/modules/signUp/index.js:11]:
```suggestion
export const createUserRequest = () => ({ type: CREATE_USER_REQUEST});
```

File: src/store/modules/signUp/index.js [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+export const signUpSuccess = (user) => ({ type: SIGNUP_SUCCESS, payload: user });
+export const signUpFailure = (error) => ({ type: SIGNUP_FAILURE, payload: error });
+
+export function* getSignupRequestSaga({ payload }) {
```
Comment (Jonathan4github) [inline src/store/modules/signUp/index.js]: getSignupRequestSaga to getSignupRequest

File: src/store/modules/signUp/signup.reducers.test.js [hunk 22]
```diff
@@ -0,0 +1,118 @@
+import { runSaga } from 'redux-saga';
+import sinon from 'sinon';
+import http from 'utils/http';
+
+import * as types from './types';
+
+// actions
+import signup, {
+  createUserRequest,
+  getSignupRequestSaga
```
Comment (Jonathan4github) [inline src/store/modules/signUp/signup.reducers.test.js]: getSignupRequestSaga to getSignupRequest

File: src/validations/Signup.js [hunk 23]
```diff
@@ -0,0 +1,38 @@
+export default {
+  Signup: (validator) => [
```
Comment (Jonathan4github) [inline src/validations/Signup.js]: export default it

**Comment by iverenshaguy** [comment]
@Jonathan4github Please bump up the test coverage on this feature. It is too low

**Comment by iverenshaguy** [comment]
Also, I can't see any e2e tests. Did you write any? If you didn't, please do that. Fill in the fields with invalid data, check that the correct errors are displayed, and then fill them with accurate data, check that no error is displayed. Do not click to submit the form though. This is because we do not want to keep signing up many users, we don't have the resource for this.

File: package.json (hunk 3 above)
Comment (successgilli) [inline package.json]: @Jonathan4github , you have not yet removed jest-mock

File: src/components/Signup/Signup.spec.jsx [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+import { shallow } from 'enzyme';
+
```
Comment (successgilli) [inline src/components/Signup/Signup.spec.jsx:4]: Please, use proper import ordering.

**Review by successgilli — CHANGES_REQUESTED** [review]
Some of the feedback I gave has not been implemented. Please, is there a reason for this ?

File: src/utils/validator/index.js [hunk 25]
```diff
@@ -0,0 +1,33 @@
+import validator from 'validator';
+
+export default class FormValidator {
```
Comment (successgilli) [inline src/utils/validator/index.js:3]: Do you think we need jsdoc for this ?

**Review by Jonathan4github — COMMENTED** [review]
> Also, I can't see any e2e tests. Did you write any? If you didn't, please do that. Fill in the fields with invalid data, check that the correct errors are displayed, and then fill them with accurate data, check that no error is displayed. Do not click to submit the form though. This is because we do not want to keep signing up many users, we don't have the resource for this.

Okay I'm on that.

File: src/store/modules/signUp/index.js (hunk 20 above)
Comment (successgilli) [reply src/store/modules/signUp/index.js:11]: This has not been updated

**Review by successgilli — APPROVED** [review]
Good job @Jonathan4github implementing all feedback.

**Review by iverenshaguy — APPROVED** [review]

**Pull request merged**